### DISCUSSION TITLE
Crash at CefSharp.Internals.ClientAdapter.OnProcessMessageReceived

### DISCUSSION THREAD

**Discussion by eadan-y** [cefsharp/CefSharp#3573] [Q&A]
Hi
This happens on version 86 as well as the latest (90)
added the libcef PDB.
This is the stack:

System.NullReferenceException: Object reference not set to an instance of an object.   at 
CefSharp.Internals.ClientAdapter.OnProcessMessageReceived(ClientAdapter* , scoped_refptr<CefBrowser>* browser, scoped_refptr<CefFrame>* frame, cef_process_id_t source_process, scoped_refptr<CefProcessMessage>* message) in c:\\projects\\cefsharp\\cefsharp.core.runtime\\internals\\clientadapter.cpp:line 1356

It happens after I hide a WPF window containing the browser. 
I am communicating between my C# application and the browser page (mostly browser calls C# methods).

I think this happens when the browser gets disposed while there is a call from javascript to C# (via JavascriptObjectRepository) still running.

**Comment by amaitland** [comment]
> This is the stack:

Which version corresponds to the stack trace?

**Reply by eadan-y** [reply]
90.6.5

**Reply by amaitland** [reply]
https://github.com/cefsharp/CefSharp/blob/cefsharp/90/CefSharp.Core.Runtime/Internals/ClientAdapter.cpp#L1356

It's not clear how that line number would cause a nullreference exception. 

Can you provide a more detailed example of how to reproduce this? 

> It happens after I hide a WPF window containing the browser.

Are you disposing of the ChromiumWebBrowser instance when you hide the window?

**Reply by eadan-y** [reply]
I am not disposing the browser manually.
I have now tried to dispose it before hiding the window, but the exception is still raised.
I'm still trying to understand the exact scenario for that exception.

Can a browser be disposed while a javascript call to c# is "hanging"? can that cause exception? Can I check from outside of CefSharp code if there is a waiting async call from javascript?

**Reply by amaitland** [reply]
> I am not disposing the browser manually.

Are you minimizing or unloading the WPF window? Performing an action that causes the parent WPF window to unload will by default cause the ChromiumWebBrowser instance to dispose, you can control this by setting http://cefsharp.github.io/api/90.6.x/html/P_CefSharp_Wpf_ChromiumWebBrowser_CleanupElement.htm to null. You will need to manually Dispose of the ChromiumWebBrowser instances. 

By default minimizing a WPF window won't unload it. Are you minimizing or doing something else? 

> Can a browser be disposed while a javascript call to c# is "hanging"?

If by hanging you mean pending then possibly, the line number doesn't match with a line that would cause a null reference exception, I'd expect an access violation if there was a error on line 1356.

Some extra null checks could be added, they would be speculative as there's not a clear example of how to reproduce the problem. 

> Can I check from outside of CefSharp code if there is a waiting async call from javascript?

You would manually  keep track of then wait on all pending tasks. Keep a list, call Task.WhenAll etc.

**Reply by amaitland** [reply]
> You would manually keep track of then wait on all pending tasks. Keep a list, call Task.WhenAll etc.

Actually,  my mistake the calls originate from JavaScript. As the calls originate from a different process there's no way I can think of to deal with this yourself.

**Reply by eadan-y** [reply]
I can try create an example app to reproduce the issue but it will take some time. 
Is there anything I can do to get more data when I see the exception on my end to help you understand better it's origin? I can reproduce it on my end but it's on a complicated app.

**Reply by amaitland** [reply]
You can build your own version from source and add extra logging/error handing 

https://github.com/cefsharp/CefSharp/tree/cefsharp/90

**Reply by amaitland** [reply]
I've made some speculative changes in commit https://github.com/cefsharp/CefSharp/commit/2e8751665b07e909e3f68e4372a0d4773ff2690f

Basically some additional null checks.

Once https://ci.appveyor.com/project/cefsharp/cefsharp/builds/39239530 is complete you should be able to test `90.6.50-RCI4046` from https://www.myget.org/feed/cefsharp/package/nuget/CefSharp.Wpf

**Reply by eadan-y** [reply]
Hey, we weren't able to figure this out yet.
Question - is there any chance we could use your help as a consultant? 
I think a few hours of your knowledge about the project probably worth days of our time. 
This is a paid job obviously. 
In case it is relevant, please contact me at eadan.y (@walkme.com)
CC @amaitland

**Reply by eadan-y** [reply]
Also, the speculative changes does seem to help (although I tested with 91.1.10-CI4048).
Is there any chance to create a hotfix for version 90 containing that fix?

**Reply by amaitland** [reply]
> Also, the speculative changes does seem to help (although I tested with 91.1.10-CI4048).

Can you clarify `help` in this context?` Are you still seeing the problem just less often or does the problem appear to be resolved?

> Question - is there any chance we could use your help as a consultant?

If the problem still reproduces after the changes I made in commit https://github.com/cefsharp/CefSharp/commit/2e8751665b07e909e3f68e4372a0d4773ff2690f then I'm happy to discuss that option.

Generally speaking I would prefer that people/organisations fund my efforts via [GitHub Sponsors](https://github.com/sponsors/amaitland)

> Is there any chance to create a hotfix for version 90 containing that fix?

I'll likely release a new version at the end of next week. If your company would like to [sponsor](https://github.com/sponsors/amaitland) me (or if someone from your company is already sponsoring) then I'd be happy to prioritise your request and release a new version `ASAP`.

**Reply by eadan-y** [reply]
The problem seems to be resolved in 91.1.10-CI4048.
I have now compiled my own version from 90 branch (that contains the new commit done after the tag) to a locally hosted nuget package to see if it fixes the issue as well, as we don't want to use the CI version. 
Will the new version (~next week) be of 91 or of 90?

As for the sponsoring - We would love to support you while maintaining this project. We're running it within the organization.

**Reply by amaitland** [reply]
> Will the new version (~next week) be of 91 or of 90?

I'll release an updated version of `90` including update to `Chromium 90.0.4430.212`. I'll release a `91-pre` release a few days after.

**Reply by amaitland** [reply]
Version `90.6.70` is now available on `Nuget.org`.

**Reply by raziele** [reply]
Hi @amaitland, nice to e-meet. I'm @eadan-y's manager.
I'd like to thank you for your donation and work. We started supporting the project via the sponsership channel here on github (I hope you can see it already).

We'd love to discuss expanding this support for the project as well as exploring additional partnership options we've been thinking of.

If you're interested, please reach out to me at raziel.einhorn(at)walkme.com and we'll set up a call.
